src/bun.js/http.exports.js

```diff
@@ -0,0 +1,482 @@
+      this.#server = undefined;
+    }
+  }
+
+  // @ts-ignore
+  listen(...args) {
+    const server = this;
+    const [options, listening_cb] = _normalizeArgs(args);
+    const res_class = this.#options.ServerResponse || ServerResponse;
+    const req_class = this.#options.IncomingMessage || IncomingMessage;
+
+    try {
+      this.#server = Bun.serve({
+        port: options.port,
+        hostname: options.host,
+
+        fetch(req) {
+          return new Promise((reply, reject) => {
+            const http_req = new req_class(req);
+            const http_res = new res_class({ reply, req: http_req });
+
+            http_req.once('error', (err) => reject(err));
+            http_res.once('error', (err) => reject(err));
+
+            server.emit('request', http_req, http_res);
+          });
+        },
+      });
+
+      if (listening_cb) listening_cb();
+    } catch (err) {
+      this.emit('error', new Error(`bun-http-polyfill: Bun.serve failed: ${err.message}`));
+    }
+  }
+}
+
+export class IncomingMessage extends Readable {
+  constructor(req) {
+    let body;
+    let body_offset = 0;
+    const rawHeaders = [];
+    const method = req.method;
+    const headers = Object.create(null);
+    const no_body = 'GET' === method || 'HEAD' === method;
+
+    for (const key of req.headers.keys()) {
+      const value = req.headers.get(key);
+
+      headers[key] = value;
+      rawHeaders.push(key, value);
+    }
+
+    const read_body = async (size) => {
+      // TODO: streaming
+      try {
+        body = Buffer.from(await req.arrayBuffer());
+        this.push(body.subarray(body_offset, body_offset += size));
+      } catch (err) {
+        this.destroy(err);
+      }
+    };
+
+    super({
+      read: size => {
+        if (no_body) {
+          this.push(null);
+          this.complete = true;
+        }
+
+        else {
+          if (!body) read_body(size);
+
+          else {
+            if (body_offset >= body.length) {
+              this.push(null);
+              this.complete = true;
+            } else {
+              this.push(body.subarray(body_offset, body_offset += size));
+            }
+          }
+        }
+      },
+    });
+
+    const url = new URL(req.url);
+    // TODO: reuse trailer object?
+    // TODO: get hostname and port from Bun.serve and calculate substring() offset
+
+    this.complete = false;
+    this.rawTrailers = [];
+    this.headers = headers;
+    this.httpVersion = '1.1';
+    this.method = req.method;
+    this.httpVersionMajor = 1;
+    this.httpVersionMinor = 1;
+    this.rawHeaders = rawHeaders;
+    this.trailers = Object.create(null);
+    this.url = url.pathname + url.search;
+  }
+
+  get aborted() {
+    throw new Error('not implemented');
+  }
+
+  get connection() {
+    throw new Error('not implemented');
+  }
+
+  get statusCode() {
+    throw new Error('not implemented');
+  }
+
+  get statusMessage() {
+    throw new Error('not implemented');
+  }
+
+  get socket() {
```

Comment:
should this be a private property with a getter so that a new EventEmitter isn't created each time?